query/dev/system/src/main/java/org/codice/ddf/admin/query/dev/system/persist/CreatePackageDependencyGraph.java

```diff
@@ -0,0 +1,126 @@
+
+  private static final String DEFAULT_GRAPH_NAME = "packageDependenciesGraph.graphml";
+  private static final String FUNCTION_NAME = "createPackageDependenciesGraph";
+  private static final String DESCRIPTION = "Saves a graph called \'" + DEFAULT_GRAPH_NAME + "\' consisting of bundles as the vertices and services as edges."
+          + "By default, the graph save path is under ddf.home. The file format is in graphml. Look here for more information: http://graphml.graphdrawing.org/";
+
+  private static final String SAVE_DIR_FIELD_NAME = "saveDirectory";
+
+  private static final BooleanField RETURN_TYPE = new BooleanField();
+  private static final Set<String> ERROR_CODES = ImmutableSet.of(DIRECTORY_DOES_NOT_EXIST);
+
+  private GraphMLExporter<BundleField, DependencyEdge<PackageField>> exporter;
+  private static final BundleGraphProvider.BundleVertexAttributeProvider BUNDLE_VERTEX_PROV = new BundleGraphProvider.BundleVertexAttributeProvider();
+  private static final PackageGraphProvider.PackageEdgeProvider PKG_EDGE_PROV = new PackageGraphProvider.PackageEdgeProvider();
+
+  // TODO: tbatie - 11/9/17 - Separate saveDir into FilePathField and add a method that allows you to trigger file path validation
+  private StringField saveDir;
+  private List<Field> arguments;
+  private BundleUtils bundleUtils;
+
+  public CreatePackageDependencyGraph(BundleUtils bundleUtils) {
+    super(FUNCTION_NAME, DESCRIPTION);
+    this.bundleUtils = bundleUtils;
+    saveDir = new StringField(SAVE_DIR_FIELD_NAME);
+    arguments = ImmutableList.of(saveDir);
+
+    exporter = new GraphMLExporter<>();
+    exporter.setVertexAttributeProvider(BUNDLE_VERTEX_PROV);
+    exporter.setEdgeAttributeProvider(PKG_EDGE_PROV);
+
+    Stream.concat(BUNDLE_VERTEX_PROV.getAttributes().stream(), PKG_EDGE_PROV.getAttributes().stream())
+            .forEach(attri -> exporter.registerAttribute(attri.getAttriName(), attri.getCategory(), attri.getType()));
+  }
+
+  @Override
+  public BooleanField performFunction() {
+    DirectedGraph<BundleField, DependencyEdge<PackageField>> graph = createPkgDependencyGraph();
+    String savePath = saveDir.getValue() == null ? System.getProperty("ddf.home") : saveDir.getValue();
+
+    try {
+      exporter.exportGraph(graph, Paths.get(savePath, DEFAULT_GRAPH_NAME).toFile());
+    } catch (ExportException e) {
+      LOGGER.error("Failed to export features graph.", e);
+      addErrorMessage(failedPersistError());
+    }
+    return new BooleanField(!containsErrorMsgs());
+  }
+
+  private DirectedGraph<BundleField, DependencyEdge<PackageField>> createPkgDependencyGraph() {
+    DirectedGraph graph = new DirectedPseudograph<>(ServiceReferenceField.class);
+    List<BundleField> bundles = bundleUtils.getAllBundleFields().getList();
+    bundles.forEach(graph::addVertex);
+
+    for(BundleField bundle : bundles) {
+      bundle
+              .importedPackages()
+              .forEach(
```

Comment:
✏️ I would definitely not try to do this inside a lambda. Either extract the lambda to a private method or switch to an iterative solution.